File: 0108_convert_sorted_array_to_binary_search_tree/solution.php
// LeetCode 0108 - Convert Sorted Array to Binary Search Tree
// https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/

class TreeNode {
    public $val = 0;
    public $left = null;
    public $right = null;
    function __construct($val = 0, $left = null, $right = null) {
        $this->val = $val;
        $this->left = $left;
        $this->right = $right;
    }
}

class Solution {
    /**
     * @param Integer[] $nums
     * @return TreeNode
     */
    function sortedArrayToBST($nums) {
        return $this->build($nums, 0, count($nums) - 1);
    }

    private function build($nums, $left, $right) {
        if ($left > $right) {
            return null;
        }
        $mid = intdiv($left + $right + 1, 2);
        $root = new TreeNode($nums[$mid]);
        $root->left = $this->build($nums, $left, $mid - 1);
        $root->right = $this->build($nums, $mid + 1, $right);
        return $root;
    }
}

File: 0108_convert_sorted_array_to_binary_search_tree/Solution.kt
// LeetCode 0108 - Convert Sorted Array to Binary Search Tree
// https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/

class TreeNode(var `val`: Int) {
    var left: TreeNode? = null
    var right: TreeNode? = null
}

class Solution {
    fun sortedArrayToBST(nums: IntArray): TreeNode? {
        return build(nums, 0, nums.size - 1)
    }

    private fun build(nums: IntArray, left: Int, right: Int): TreeNode? {
        if (left > right) {
            return null
        }
        val mid = (left + right + 1) / 2
        val root = TreeNode(nums[mid])
        root.left = build(nums, left, mid - 1)
        root.right = build(nums, mid + 1, right)
        return root
    }
}
